include/seqan3/alignment/matrix/alignment_trace_matrix.hpp

```diff
@@ -0,0 +1,333 @@
+ * \tparam    query_t        The type of the query sequence.
+ * \tparam    trace_matrix_t The type of the trace matrix.
+ * \param[in] database       The database sequence.
+ * \param[in] query          The query sequence.
+ * \param[in] matrix         The trace matrix.
+ */
+template <
+    typename database_t,
+    typename query_t,
+    typename trace_matrix_t,
+    typename gapped_database_t = gapped<std::decay_t<decltype(std::declval<database_t>()[0])>>,
+    typename gapped_query_t = gapped<std::decay_t<decltype(std::declval<query_t>()[0])>>>
+//!\cond
+    requires matrix_concept<remove_cvref_t<trace_matrix_t>> &&
+             std::Same<typename remove_cvref_t<trace_matrix_t>::entry_type, trace_directions>
+//!\endcond
+inline std::pair<std::list<gapped_database_t>, std::list<gapped_query_t>>
+alignment_trace(database_t && database, query_t && query, trace_matrix_t && matrix)
+{
+    constexpr auto N = trace_directions::none;
+    constexpr auto D = trace_directions::diagonal;
+    constexpr auto L = trace_directions::left;
+    constexpr auto U = trace_directions::up;
+    int row = matrix.rows()-1;
+    int col = matrix.cols()-1;
+
+    std::list<gapped_database_t> gapped_database{};
+    std::list<gapped_query_t> gapped_query{};
+
+    if (matrix.at(0, 0) != N)
+        throw std::runtime_error{"End trace must be NONE"};
+
+    while(row >= 0 && col >= 0)
+    {
+        trace_directions dir = matrix.at(row, col);
+        if ((dir & L) == L)
+        {
+            col = std::max<int>(col - 1, 0);
+            gapped_database.push_front(database[col]);
+            gapped_query.push_front(gap::GAP);
+        }
+        else if ((dir & U) == U)
+        {
+            row = std::max<int>(row - 1, 0);
+            gapped_database.push_front(gap::GAP);
+            gapped_query.push_front(query[row]);
+        }
+        else if ((dir & D) == D)
+        {
+            row = std::max<int>(row - 1, 0);
+            col = std::max<int>(col - 1, 0);
+            gapped_database.push_front(database[col]);
+            gapped_query.push_front(query[row]);
+        }
+        else
+        {
+            if (row == 0 || col == 0)
+                break;
+            throw std::runtime_error{"Trace not found"};
+        }
+    }
+    return {gapped_database, gapped_query};
+}
+
+/*!\brief A trace matrix represented in a one-dimensional std::vector
+ * \ingroup alignment_matrix
+ *
+ * \details
+ *
+ * This data structure stores the matrix in a flat way using the
+ * std::vector<#entry_type> data structure where each row is stored
+ * continuously.
+ *
+ * ## Example
+ *
+ * \snippet test/snippet/alignment/matrix/alignment_trace_matrix_vector.cpp code
+ *
+ * ### Output
+ * \include test/snippet/alignment/matrix/alignment_trace_matrix_vector.out
+ */
+template <>
+struct alignment_trace_matrix<std::vector<trace_directions>>
+    : public rowwise_matrix<trace_directions>
+{
+    using rowwise_matrix<trace_directions>::rowwise_matrix;
```

Comment:
are alignment computed row-wise or column-wise in our DP-code? This should match the matrix type...

Reply:
This will likely change to column-wise, but this will be addressed later.